File: backend/common/common/types/ai_analysis.py
"""
Shared type definitions for AI analysis responses.
These types ensure consistency between frontend and backend.
"""
from pydantic import BaseModel, Field
from typing import List, Optional, Literal

class Finding(BaseModel):
    """A single clinical finding from AI analysis"""
    id: Optional[str] = Field(None, description="Unique identifier for this finding")
    type: Literal["normal", "abnormal", "suspicious"] = Field(
        ..., description="Classification of the finding"
    )
    severity: Literal["none", "mild", "moderate", "severe", "critical"] = Field(
        ..., description="Severity level of the finding"
    )
    category: Literal[
        "lung_parenchyma",
        "mediastinum",
        "bones",
        "soft_tissue",
        "airways",
        "cardiovascular",
        "other"
    ] = Field(..., description="Anatomical category")
    title: str = Field(..., description="Short title for the finding")
    description: str = Field(..., description="Detailed description")
    confidence: float = Field(..., ge=0.0, le=1.0, description="Confidence score")
    slice_locations: Optional[List[int]] = Field(
        None, description="List of slice numbers where finding appears"
    )
    supporting_evidence: Optional[List[str]] = Field(
        None, description="Supporting observations"
    )


class Recommendation(BaseModel):
    """A clinical recommendation based on findings"""
    id: Optional[str] = Field(None, description="Unique identifier")
    priority: Literal["urgent", "high", "routine", "low"] = Field(
        ..., description="Priority level"
    )
    category: Literal[
        "follow_up",
        "intervention",
        "consultation",
        "additional_imaging",
        "monitoring"
    ] = Field(..., description="Type of recommendation")
    title: str = Field(..., description="Short title")
    description: str = Field(..., description="Detailed description")
    urgency: Literal["immediate", "urgent", "routine", "elective"] = Field(
        ..., description="Urgency level"
    )
    timeframe: Optional[str] = Field(
        None, description="Suggested timeframe (e.g., '3_months', '1_week')"
    )
    rationale: str = Field(..., description="Reason for recommendation")


class QualityAssessment(BaseModel):
    """Image quality assessment"""
    score: float = Field(..., ge=0.0, le=1.0, description="Overall quality score")
    issues: List[str] = Field(default_factory=list, description="Quality issues found")


class SliceAnalysisMetadata(BaseModel):
    """Metadata for single slice analysis"""
    slice_number: int = Field(..., description="Current slice number")
    total_slices: int = Field(..., description="Total number of slices")
    anatomical_region: str = Field(..., description="Anatomical region name")
    timestamp: str = Field(..., description="ISO format timestamp")
    model_version: str = Field(default="mock", description="AI model version")
    processing_time_ms: int = Field(..., description="Processing time in milliseconds")


class SliceAnalysisResponse(BaseModel):
    """Response for single slice analysis"""
    analysis_type: Literal["single_slice"] = "single_slice"
    metadata: SliceAnalysisMetadata
    quality_assessment: QualityAssessment
    findings: List[Finding]
    summary: str = Field(..., description="Brief overall summary")


class SliceRange(BaseModel):
    """Range of slices analyzed"""
    start: int = Field(..., description="Starting slice number")
    end: int = Field(..., description="Ending slice number")
    total_analyzed: int = Field(..., description="Total slices analyzed")


class BatchAnalysisMetadata(BaseModel):
    """Metadata for batch analysis"""
    patient_id: str = Field(..., description="Patient identifier")
    file_id: str = Field(..., description="File identifier")
    file_name: str = Field(..., description="File name")
    slice_range: SliceRange
    timestamp: str = Field(..., description="ISO format timestamp")
    model_version: str = Field(default="mock", description="AI model version")
    processing_time_ms: int = Field(..., description="Processing time in milliseconds")


class OverallSummary(BaseModel):
    """Overall summary of batch analysis"""
    title: str = Field(..., description="Summary title")
    content: str = Field(..., description="Summary content")
    confidence: float = Field(..., ge=0.0, le=1.0, description="Overall confidence")
    urgency: Literal["immediate", "urgent", "routine", "elective"] = Field(
        ..., description="Overall urgency level"
    )


class BatchAnalysisResponse(BaseModel):
    """Response for batch analysis of multiple slices"""
    analysis_type: Literal["batch_analysis"] = "batch_analysis"
    metadata: BatchAnalysisMetadata
    overall_summary: OverallSummary
    findings: List[Finding]
    recommendations: List[Recommendation]
    differential_diagnosis: Optional[List[str]] = Field(
        default_factory=list, description="Possible diagnoses"
    )


class SliceAnalysisRequest(BaseModel):
    """Request for single slice analysis"""
    patient_id: str
    file_id: str
    slice_number: int
    image_data: Optional[str] = Field(
        None, description="Base64 encoded image data (optional if using file_id)"
    )


class BatchAnalysisRequest(BaseModel):
    """Request for batch analysis"""
    patient_id: str
    file_id: str
    slice_start: int
    slice_end: int
    step_size: int
    image_slices: List[str]

File: frontend/src/types/ai-analysis.ts
/**
 * TypeScript type definitions for AI Analysis
 * These types match the backend Python Pydantic models
 */

// ============================================================================
// Finding Types
// ============================================================================

export type FindingType = 'normal' | 'abnormal' | 'suspicious';
export type SeverityLevel = 'none' | 'mild' | 'moderate' | 'severe' | 'critical';
export type FindingCategory =
  | 'lung_parenchyma'
  | 'mediastinum'
  | 'bones'
  | 'soft_tissue'
  | 'airways'
  | 'cardiovascular'
  | 'other';

export interface Finding {
  id: string;
  type: FindingType;
  severity: SeverityLevel;
  category: FindingCategory;
  title: string;
  description: string;
  confidence: number; // 0.0 to 1.0
  slice_locations?: number[];
  supporting_evidence?: string[];
  location?: {
    region?: string;
    side?: string;
    coordinates?: any;
  };
}

// ============================================================================
// Recommendation Types
// ============================================================================

export type RecommendationPriority = 'urgent' | 'high' | 'routine' | 'low';
export type RecommendationCategory =
  | 'follow_up'
  | 'intervention'
  | 'consultation'
  | 'additional_imaging'
  | 'monitoring';
export type UrgencyLevel = 'immediate' | 'urgent' | 'routine' | 'elective';

export interface Recommendation {
  id: string;
  priority: RecommendationPriority;
  category: RecommendationCategory;
  title: string;
  description: string;
  urgency: UrgencyLevel;
  timeframe?: string;
  rationale: string;
}

// ============================================================================
// Slice Analysis (Single Slice)
// ============================================================================

export interface QualityAssessment {
  score: number; // 0.0 to 1.0
  issues: string[];
}

export interface SliceAnalysisMetadata {
  slice_number: number;
  total_slices: number;
  anatomical_region: string;
  timestamp: string; // ISO format
  model_version: string;
  processing_time_ms: number;
}

export interface SliceAnalysisResponse {
  analysis_type: 'single_slice';
  metadata: SliceAnalysisMetadata;
  quality_assessment: QualityAssessment;
  findings: Finding[];
  summary: string;
}

// ============================================================================
// Batch Analysis (Multiple Slices)
// ============================================================================

export interface SliceRange {
  start: number;
  end: number;
  total_analyzed: number;
}

export interface BatchAnalysisMetadata {
  patient_id: string;
  file_id: string;
  file_name: string;
  slice_range: SliceRange;
  timestamp: string;
  model_version: string;
  processing_time_ms: number;
}

export interface OverallSummary {
  title: string;
  content: string;
  confidence: number; // 0.0 to 1.0
  urgency: UrgencyLevel;
}

export interface BatchAnalysisResponse {
  analysis_type: 'batch_analysis';
  metadata: BatchAnalysisMetadata;
  overall_summary: OverallSummary;
  findings: Finding[];
  recommendations: Recommendation[];
  differential_diagnosis?: string[];
  measurements?: any[];
  comparison_to_prior?: any;
}

// ============================================================================
// Request Types
// ============================================================================

export interface SliceAnalysisRequest {
  patient_id: string;
  file_id: string;
  slice_number: number;
  image_data?: string; // Base64 encoded
}

export interface BatchAnalysisRequest {
  patient_id: string;
  file_id: string;
  slice_start: number;
  slice_end: number;
  step_size: number;
  image_slices: string[];
}

// ============================================================================
// UI Helper Types
// ============================================================================

export interface FindingWithUI extends Finding {
  // UI-specific additions
  icon?: string;
  color?: string;
  expanded?: boolean;
}

export interface RecommendationWithUI extends Recommendation {
  // UI-specific additions
  icon?: string;
  color?: string;
  completed?: boolean;
}

// Helper function to get finding icon based on type and severity
export function getFindingIcon(finding: Finding): string {
  if (finding.type === 'normal') return '✓';
  if (finding.severity === 'critical') return '🚨';
  if (finding.severity === 'severe') return '⚠️';
  if (finding.type === 'suspicious') return '🔍';
  return '📋';
}

// Helper function to get finding color based on severity
export function getFindingColor(finding: Finding): string {
  switch (finding.severity) {
    case 'critical': return '#dc2626'; // red-600
    case 'severe': return '#ea580c'; // orange-600
    case 'moderate': return '#f59e0b'; // amber-500
    case 'mild': return '#3b82f6'; // blue-500
    case 'none': return '#10b981'; // green-500
    default: return '#6b7280'; // gray-500
  }
}

// Helper function to get recommendation icon
export function getRecommendationIcon(rec: Recommendation): string {
  switch (rec.category) {
    case 'follow_up': return '📅';
    case 'intervention': return '🏥';
    case 'consultation': return '👨‍⚕️';
    case 'additional_imaging': return '🔬';
    case 'monitoring': return '👁️';
    default: return '💡';
  }
}

// Helper function to get recommendation color
export function getRecommendationColor(rec: Recommendation): string {
  switch (rec.priority) {
    case 'urgent': return '#dc2626';
    case 'high': return '#f59e0b';
    case 'routine': return '#3b82f6';
    case 'low': return '#10b981';
    default: return '#6b7280';
  }
}
